File: src/hand.cpp
// File:     hand.cpp
// Location: src
// Purpose:  implement Hand class
// Author:   Stephen Gold sgold@sonic.net
// (c) Copyright 2012 Stephen Gold
// Distributed under the terms of theGNU General Public License

/*
This file is part of the Gold Tile Game.

The Gold Tile Game is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by the 
Free Software Foundation, either version 3 of the License, or (at your 
option) any later version.

The Gold Tile Game is distributed in the hope that it will be useful, but 
WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY 
or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License 
for more details.

You should have received a copy of the GNU General Public License
along with the Gold Tile Game.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <iostream>
#include "network.hpp"
#include "partial.hpp"


// lifecycle

Hand::Hand(
	String const& rHandName,
	HandOpt const& rOptions,
	Socket const& rClientSocket)
:
    mName(rHandName),
	mOptions(rOptions)
{
	mClockRunningFlag = false;
	mMilliseconds = 0L;
	if (rClientSocket.IsValid() && Address(rOptions) == rClientSocket.Peer()) {
		mSocket = rClientSocket;
	}
	// mStartTime gets initialized in StartClock()
	Restart();
}

// The compiler-generated copy constructor is fine.
// The compiler-generated destructor is fine.


// operators

// The compiler-generated assignment method is fine.

bool Hand::operator==(Hand const& rOther) const {
	bool const result = (Name() == rOther.Name());

	return result;
}

Hand::operator Address(void) const {
	Address const result = mOptions;

	return result;
}

Hand::operator HandOpt(void) const {
	return mOptions;
}

Hand::operator Socket(void) const {
	return mSocket;
}

Hand::operator Tiles(void) const {
    return mTiles;
}


// misc methods

void Hand::AddScore(ScoreType points) {
    ASSERT(!IsClockRunning());

	ScoreType const new_score = mScore + points;
	ASSERT(new_score >= mScore); // check for wraparound
	mScore = new_score;
}

void Hand::AddTiles(Tiles const& rTiles) {
    mTiles.Merge(rTiles);
}

Move Hand::ChooseMove(unsigned mustPlay) const {
    ASSERT(IsClockRunning());
    ASSERT(IsLocalUser());
    
    DescribeTiles();
	Move result;
    result.GetUserChoice(mTiles, mustPlay);
    
	return result;
}

void Hand::ConnectToServer(Game& rGame) {
	ASSERT(IsRemote());
	ASSERT(rGame.AmClient());

    Address const server_address = *this;
	if (!rGame.IsConnectedToServer(server_address)) {
		Network::ConnectToServer(server_address, rGame);
	}
}

void Hand::DescribeName(void) const {
    std::cout << Name();
}

void Hand::DescribeScore(void) const {
	DescribeName();
    std::cout << " has " << ::plural(mScore, "point") << "." << std::endl;
}

void Hand::DescribeTiles(void) const {
	DescribeName();
	unsigned const count = mTiles.Count();
	std::cout << " holds " << ::plural(count, "tile") << ": " 
		      << mTiles.Description() << "." << std::endl;
}

Move Hand::GetAutomaticMove(Game const& rGame) const {
	double skip_probability = SkipProbability();
	if (rGame.MustPlay() > 0) {
		skip_probability = 0.0;
	}
    Partial partial(&rGame, HINT_NONE, skip_probability);
    partial.Suggest();
    Move result = partial.GetMove(false);

    DescribeName();
	std::cout << " " << result.Description() << "." << std::endl;

	return result;
}

Move Hand::GetRemoteMove(void) {
    Address const address = *this;
	String const move_string = mSocket.GetLine();
	Move const result = Move(move_string, true);

    DescribeName();
	std::cout << " " << result.Description() << "." << std::endl;

	return result;
}

// find the longest run of compatible tiles in the hand
Tiles Hand::LongestRun(void) const {
	Tiles const result = mTiles.LongestRun();
            
	return result;
}

// read the hand's elapsed time clock
// and if the clock is running, update mTurnMilliseconds
MsecIntervalType Hand::Milliseconds(void) const {
	MsecIntervalType result = mMilliseconds;

    if (mClockRunningFlag) {
	    MsecIntervalType const now = ::milliseconds();
		ASSERT(now >= mStartTime);
		MsecIntervalType const turn_msec = now - mStartTime;
        result += turn_msec;
		ASSERT(result >= mMilliseconds);
	}

	return result;
}

String Hand::Name(void) const {
    return mName;
}

String Hand::PlayerName(void) const {
    String result = mOptions.PlayerName();

	return result;
}

Tiles Hand::PullRandomTiles(unsigned tileCount, Tiles& rBag) {
    Tiles result;
    result.PullRandomTiles(tileCount, rBag);

	unsigned const count = result.Count();
	if (count > 0) {
	    std::cout << Name() << " drew " << plural(count, "tile") 
		     << " from the stock bag." << std::endl;
	}

    mTiles.Merge(result);

	return result;
}

void Hand::RemoveTile(Tile const& rTile) {
	mTiles.Remove(rTile);
}

void Hand::RemoveTiles(Tiles const& rTiles) {
	mTiles.Purge(rTiles);
}

void Hand::Resign(Tiles& rBag) {
    ASSERT(!IsClockRunning());
	ASSERT(!HasResigned());

	rBag.Merge(mTiles);
	mTiles.MakeEmpty();
	mResignedFlag = true;

	ASSERT(HasResigned());
}

void Hand::Restart(void) {
	ASSERT(!IsClockRunning());

	mResignedFlag = false;
	mScore = 0;
	mTiles.MakeEmpty();
}

// total points this hand has scored
ScoreType Hand::Score(void) const {
    return mScore;
}

// how much time this hand has spent, in seconds
unsigned Hand::Seconds(void) const {
	MsecIntervalType const msecs = Milliseconds();
    unsigned const result = unsigned(msecs/MSECS_PER_SECOND);

	return result;
}

void Hand::SetSocket(Socket const& rSocket) {
    ASSERT(!mSocket.IsValid());
    ASSERT(rSocket.IsValid());

	mSocket = rSocket;

    ASSERT(rSocket == mSocket);
}

double Hand::SkipProbability(void) const {
	double result = mOptions.SkipProbability();

	return result;
}

void Hand::StartClock(void) {
	ASSERT(!mClockRunningFlag);

	mStartTime = ::milliseconds();
	mClockRunningFlag = true;
}

unsigned Hand::StopClock(void) {
	ASSERT(IsClockRunning());

	mMilliseconds = Milliseconds();
	mClockRunningFlag = false;

	// return the number of elapsed seconds
	unsigned const result = unsigned(mMilliseconds/MSECS_PER_SECOND);

	return result;
}

void Hand::SubtractScore(ScoreType points) {
	ASSERT(!HasResigned());
    ASSERT(!IsClockRunning());

	ScoreType const new_score = mScore - points;
	ASSERT(new_score <= mScore); // check for wraparound
	mScore = new_score;
}

void Hand::Unresign(Tiles& rBag, Tiles const& rHand) {
	ASSERT(HasResigned());
    ASSERT(!IsClockRunning());

	rBag.Purge(rHand);
	mTiles = rHand;
	mResignedFlag = false;

	ASSERT(!HasResigned());
}


// inquiry methods

bool Hand::HasGoneOut(void) const {
    bool const result = IsEmpty() && !HasResigned();
	
	return result;
}

bool Hand::HasResigned(void) const {
    return mResignedFlag;
}

bool Hand::IsAutomatic(void) const {
    bool const result = mOptions.IsAutomatic();

	return result;
}

bool Hand::IsClockRunning(void) const {
	return mClockRunningFlag;
}

bool Hand::IsEmpty(void) const {
     bool const result = mTiles.IsEmpty();
     
     return result;
}

bool Hand::IsLocalUser(void) const {
    bool const result = mOptions.IsLocalUser();

	return result;
}

bool Hand::IsRemote(void) const {
	bool const result = mOptions.IsRemote();

	return result;
}

File: Java/GoldTile/src/goldtile/Hand.java
// File:     Hand.java
// Location: Java/GoldTile/src/goldtile
// Purpose:  Hand class for the Gold Tile Game
/**
 * @author Stephen Gold
 */
// (c) Copyright 2012 Stephen Gold
// Distributed under the terms of the GNU General Public License

/*
This file is part of the Gold Tile Game.

The Gold Tile Game is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by the
Free Software Foundation, either version 3 of the License, or (at your
option) any later version.

The Gold Tile Game is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
for more details.

You should have received a copy of the GNU General Public License
along with the Gold Tile Game.  If not, see <http://www.gnu.org/licenses/>.
*/

package goldtile;

public class Hand
    implements ReadHand
{
    // per-instance fields
    private boolean clockRunning = false;
    private boolean resigned = false;
    final private ReadHandOpt opt;    // immutable
    private int score = 0;
    private long milliseconds = 0;    // total time charged to the hand
    private long startTime = 0;
    final private String name;        // unique name for the hand
    private Tiles contents = new Tiles();

    // constructors

    public Hand(String name, ReadHandOpt opt) {
        assert name != null;
        assert opt != null;

        this.opt = new HandOpt(opt);
        this.name = name;
    }

    // methods, sorted by name

    public void addScore(int pointCount) {
        assert !isClockRunning();
        assert pointCount >= 0 : pointCount;

        final int newScore = score + pointCount;
        assert newScore >= score : score; // check for wraparound
        score = newScore;
    }

    public void addTiles(Tiles tiles) {
        assert contents != null;
        assert tiles != null;

        contents.addAll(tiles);
    }

    @Override
    public ReadMove chooseMoveAutomatic() {
        assert Game.haveInstance();
        assert GoldTile.control == Display.CONSOLE;

        Partial partial;
        try {
            partial = Partial.findBestMove(opt.getSkipProbability());
        } catch (InterruptedException exception) {
            // Interrupts should be possible only under GUI control.
            throw new AssertionError();
        }

        final ReadMove result = partial.getMove(Partial.Active.INCLUDED);
        Console.printf(String.format("%s %s.\n", name, result.describe()));

        return result;
    }

    @Override
    public ReadMove chooseMoveConsole(int mustPlay) {
        assert mustPlay >= 0;
        assert isClockRunning();
        assert opt.isLocalUser();

        final String description = describeContents();
        Console.print(description);

        return Move.chooseConsole(contents, mustPlay);
    }

    @Override
    public ReadMove chooseMoveRemote() {
        // TODO
        return null;
    }

    @Override
    public Tiles copyContents() {
        return new Tiles(contents);
    }

    @Override
    public int countContents() {
        return contents.size();
    }

    @Override
    public String describeContents() {
        final int count = contents.size();

        return String.format("%s holds %s: %s.\n",
                name,
                StringExt.plural(count, "tile"),
                contents.describe());
    }

    @Override
    public String describeScore() {
        return String.format("%s has %s.\n",
                name,
                StringExt.plural(score, "point"));
    }

    @Override
    public Tiles findLongestRun() {
        return contents.findLongestRun();
    }

    /**
     * Read the hand's elapsed time clock, and if the clock
     * is running, update the milliseconds field.
     *
     * @return
     */
    @Override
    public long getMillisecondsUsed() {
        long result = milliseconds;

        if (clockRunning) {
            final long now = System.currentTimeMillis();
            assert now >= startTime;
            final long turnMsec = now - startTime;
            result += turnMsec;
            assert result >= milliseconds;
        }

        return result;
    }

    @Override
    public String getName() {
        return name;
    }

    @Override
    public ReadHandOpt getOpt() {
        return opt;
    }

    @Override
    public int getSecondsUsed() {
        return (int)(getMillisecondsUsed()/Global.MILLISECONDS_PER_SECOND);
    }

    @Override
    public int getScore() {
        return score;
    }

    @Override
    public boolean hasGoneOut() {
        return contents.size() == 0 && !resigned;
    }

    @Override
    public boolean hasResigned() {
        return resigned;
    }

    @Override
    public boolean isClockRunning() {
        return clockRunning;
    }

    @Override
    public boolean isDisconnected() {
        return false; // TODO
    }

    public void removeTiles(Tiles tiles) {
        contents.removeAll(tiles);
    }

    public Tiles resign() {
        assert !isClockRunning();
        assert !hasResigned();

        final Tiles tiles = contents;
        contents = new Tiles();
        resigned = true;

        assert hasResigned();
        return tiles;
    }

    public void startClock() {
        assert !clockRunning;

        startTime = System.currentTimeMillis();
        clockRunning = true;
    }

    public void stopClock() {
        assert clockRunning;

        milliseconds = System.currentTimeMillis() - startTime;
        clockRunning = false;
    }
}
